src/ast/passes/semantic_analyser.cpp

```diff
@@ -1132,6 +1137,26 @@ void SemanticAnalyser::visit(Call &call)
 
     call.type = CreateUInt64();
   }
+  else if (call.func == "offsetof")
+  {
+    check_nargs(call, 2);
+    auto &element_arg = *call.vargs->at(1);
+    String &element = static_cast<String&>(element_arg);
```

Comment:
> I don't think that this should be a `String` as it'd have to be quoted. Struct field names should follow the rules for identifiers, so I think that the proper thing to do here is to `dynamic_cast` the first arg to `Identifier *` and then check if it hasn't returned `nullptr` (which would also cover the case then `f1.f2` is used).

I try

```suggestion
--- a/src/ast/passes/codegen_llvm.cpp
+++ b/src/ast/passes/codegen_llvm.cpp
@@ -1029,8 +1029,8 @@ void CodegenLLVM::visit(Call &call)
   else if (call.func == "offsetof")
   {
     auto &element_arg = *call.vargs->at(1);
-    String &element = static_cast<String&>(element_arg);
-    auto &field = call.vargs->at(0)->type.GetField(element.str);
+    Identifier &element = dynamic_cast<Identifier&>(element_arg);
+    auto &field = call.vargs->at(0)->type.GetField(element.ident);
     expr_ = b_.getInt64(field.offset);
   }
   else if (call.func == "strncmp") {
```

 but get error:

```
ERROR: Failed to compile: std::bad_cast
```

Reply:
I'd try:
```cpp
auto element = dynamic_cast<Identifier *>(call.vargs->at(1));
if (element == nullptr)
   ...
```
notice using a pointer (which allows to check for NULL) instead of a reference.